### ISSUE TITLE
Instantiating first Realm causes memory leak originating in Swift reflection code

### ISSUE THREAD

**Issue by jviney** [realm/realm-cocoa#2933]
I've started trying to fix memory leaks on an app, and noticed that there seems to be a leak when instantiating a Realm.

Here's example project where I've included a few screenshots and the Instruments trace showing the problem.

https://github.com/jviney/realm-memory-leak-test

[Screenshot 1](https://github.com/jviney/realm-memory-leak-test/blob/master/Leaks%20Screenshot%201.png) [Screenshot 2](https://github.com/jviney/realm-memory-leak-test/blob/master/Leaks%20Screenshot%202.png) [Screenshot 3](https://github.com/jviney/realm-memory-leak-test/blob/master/Leaks%20Screenshot%203.png)

The number of objects leaked correlates with the number of properties your Realm objects have. If you comment out the definition of the `Person` class the leak is smaller.

Thanks in advance.

**Comment by vtongcoder** [comment]
I also have this leak on my app, cause crash on some users' devices.
Any solution to fix?
Thanks a lot

**Comment by bdash** [comment]
Thanks for the reproducible test case, @jviney!

The leaked `_NSContiguousString` instances never appear to escape from the Swift runtime, which makes that leak feel like a bug within the Swift runtime.

The leaked `_NativeDictionaryStorageOwner` relates to the `listProperties` member of Realm's `DynamicObject` class, but no `DynamicObject` instances are leaked. Most of the retain / release traffic on the leaked instance, including the mismatched retain, is within Swift's reflection machinery.

@vtongcoder These leaks happen during schema initialization, which is typically a one-time event during application initialization, and therefore shouldn't significantly impact the memory usage of your application. It's hard to see how this leak could relate to any crash you're seeing. It'd be great if you could file a separate GitHub issue about your problem.

**Comment by jviney** [comment]
Small world @bdash ! You're working at Realm now :+1: ! Thanks for looking into this issue.

Agreed that these leaks are pretty small and shouldn't significantly increase an app's memory usage.

**Comment by zubko** [comment]
I agree that it's not such a big leak, but anyways, it would be great to have 0 leaks in the app, so I'm reporting that the leaks during initialization are still a case (as of 0.98 Swift):

(Screenshot 1)[https://dl.dropboxusercontent.com/u/14785241/Screen%20Shot%202016-02-06%20at%202.57.14%20AM.png]
(Screenshot 2)[https://dl.dropboxusercontent.com/u/14785241/Screen%20Shot%202016-02-06%20at%202.57.20%20AM.png]

**Comment by iantheparker** [comment]
YES! It's not just me! All of these leaks in my screenshot are from my Realm init.

<img width="1280" alt="screen shot 2016-02-09 at 1 41 41 pm" src="https://cloud.githubusercontent.com/assets/1368966/12926792/005ef878-cf33-11e5-9600-8b3703974b56.png">

**Comment by annicaburns** [comment]
We are seeing the same thing on the current version of SwiftRealm and would love to get this cleaned up.

**Comment by i-schuetz** [comment]
Also here, always getting a leak in `Realm()`

**Comment by fewspider** [comment]
some problem

**Comment by SaidSikira** [comment]
v0.98.8 still getting a leak when Realm object is initialized

**Comment by jpsim** [comment]
Lots of people have commented here, but has anyone filed a radar or Swift bug? Based on @bdash's earlier comment, this appears to be a leak in the Swift runtime, which Realm Swift happens to trigger:

> The leaked `_NSContiguousString` instances never appear to escape from the Swift runtime, which makes that leak feel like a bug within the Swift runtime.

**Comment by bdash** [comment]
I've retitled this to make it clear that this is about a one-off leak originating in Swift's reflection machinery that occurs when initializing the shared schema. As best as I can tell, this is due to a bug in the Swift runtime.

We've also had recent reports of a leak related to `RealmCoordinator::get_realm` which is being tracked in #3270. Though I've not spent any time analyzing this second report, it appears at first glance to be a bug on our end.

**Comment by tgoyne** [comment]
I did verify that #3270 is a bug on our end yesterday (the initial refcount is wrong so it never gets freed, but now I need to verify that everything is correct with them actually being destroyed).

**Comment by jpsim** [comment]
Whatever the cause, this leak still seems to be happening with Realm 0.103.1 and Xcode 7.3.1:

![image](https://cloud.githubusercontent.com/assets/474794/15444385/d9d2356e-1ea4-11e6-8022-3a6bfd081033.png)

**Comment by CBeloch** [comment]
Looks like this leak still exists in 1.0.2.
It doesn't leak when I use unpersisted records

![bildschirmfoto 2016-07-28 um 14 40 52](https://cloud.githubusercontent.com/assets/171119/17212778/7b1c4a7e-54d1-11e6-84b8-e590fa5cf6aa.png)

**Comment by zhenjiangliu24** [comment]
same memory leak for the _NSContiguousString

**Comment by mjohnson12** [comment]
I'm getting the same _NSContiguousString leak as well in 1.1.0.

**Comment by gabovanlugo** [comment]
This is my capture:
<img width="1552" alt="screen shot 2016-09-28 at 9 18 28 am" src="https://cloud.githubusercontent.com/assets/1279607/18917192/a733255e-855c-11e6-84d8-65ed0994fe2c.png">

**Comment by jpsim** [comment]
Any response from Apple on this @zhenjiangliu24 @mjohnson12 @gabovanlugo? Have any of you filed radars or tickets on https://bugs.swift.org that we can dupe or follow?

**Comment by wongzigii** [comment]
Any update for this thread?

**Comment by demensdeum** [comment]
Any workaround?

**Comment by bdash** [comment]
The leak as reported by @jviney (and demonstrated in [his test project](https://github.com/jviney/realm-memory-leak-test)) appears to be fixed as of Xcode 8.3 with Swift 3.1. I've shared the updated project I tested with at https://github.com/bdash/realm-memory-leak-test.

If anyone is seeing leaks of any kind with Xcode 8.3 and Swift 3.1, it is very likely to be a different issue than was originally reported here. You'll need to file new issues about what you're seeing since the problem this GitHub issue was tracking has been fixed.
